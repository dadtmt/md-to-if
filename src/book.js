import * as R from 'ramda'
import { snakeCase } from 'change-case'

export const getSceneName = R.pipe(
  R.propOr([], 'content'),
  R.head,
  R.propOr('unnamed', 'content'),
  snakeCase
)

export const splitByScene = R.pipe(
  R.converge(R.prepend, [
    R.head,
    R.pipe(
      R.tail,
      R.splitWhen(
        R.where({
          type: R.equals('heading'),
          level: R.equals(2),
        })
      )
    ),
  ]),
  R.zipObj(['heading', 'content', 'sourceLeft']),
  ({ heading, content, sourceLeft }) => ({
    scene: {
      name: getSceneName(heading),
      content: [heading, ...content],
    },
    sourceLeft,
  })
)

const book = (source, scenes = []) => {
  const { scene, sourceLeft } = splitByScene(source)
  return R.isEmpty(sourceLeft)
    ? [...scenes, scene]
    : book(sourceLeft, [...scenes, scene])
}

export default book
