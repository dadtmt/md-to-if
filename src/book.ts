import * as R from 'ramda'
import { snakeCase } from 'change-case'

// Content -> String
export const getSceneName = R.pipe(
  R.propOr([], 'content'),
  R.head,
  R.propOr('unnamed', 'content'),
  snakeCase
)

// int -> [Content] -> [[Content],[Content]]
const splitByHeading = level =>
  R.splitWhen(
    R.where({
      type: R.equals('heading'),
      level: R.equals(level),
    })
  )

const splitActions = (level, contentList, actionList = []) => {
  const [headOfContent, ...tailOfContent] = contentList
  const [actionContent, restOfContent] = splitByHeading(level)(tailOfContent)
  const { content, actions } =
    level <= 6
      ? splitContentAndActions(level + 1)(actionContent)
      : {
          content: [],
          actions: [],
        }
  return !R.isEmpty(restOfContent)
    ? splitActions(level, restOfContent, [
        ...actionList,
        {
          name: getSceneName(headOfContent),
          sceneContent: [headOfContent, ...content],
          actions,
        },
      ])
    : actionList
}

// [Content] -> { content: [Content], actions: [Scene] }
export const splitContentAndActions = (level: number) => contentAndActions => {
  const [content, actions] = splitByHeading(level)(contentAndActions)
  return { content, actions: splitActions(level, actions) }
}

// [Content] -> { scene: Scene, content: [Content] }
export const splitByScene = (level: number): Function =>
  R.pipe(
    R.converge(R.prepend, [R.head, R.pipe(R.tail, splitByHeading(level))]),
    R.zipObj(['heading', 'content', 'sourceLeft']),
    ({ content, ...rest }) => ({
      content,
      ...rest,
    }),
    ({ heading, content, sourceLeft }) => ({
      scene: {
        name: getSceneName(heading),
        sceneContent: [heading, ...content],
      },
      sourceLeft,
    })
  )

// [Content], [Scene] -> [Scene]
const book = (source, scenes = []) => {
  const { scene, sourceLeft } = splitByScene(2)(source)
  return R.isEmpty(sourceLeft)
    ? [...scenes, scene]
    : book(sourceLeft, [...scenes, scene])
}

export default book
