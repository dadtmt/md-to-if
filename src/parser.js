import SimpleMarkdown from 'simple-markdown'

// String -> [String]
export const matchBraces = source => /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = source => /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = source => /^\|([\s\S]*?)\]/.exec(source)

const command = {
  order: 0,
  match: matchBraces,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

// String -> [Content]
const parser = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent,
})

export default parser
