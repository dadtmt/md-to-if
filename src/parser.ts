import SimpleMarkdown, { SingleASTNode } from 'simple-markdown'

// String -> [String]
export const matchBraces = (source: string): RegExpExecArray | null =>
  /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = (source: string): RegExpExecArray | null =>
  /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = (source: string): RegExpExecArray | null =>
  /^\|([\s\S]*?)\]/.exec(source)

const parse: (capture: any, parse: any, state: any) => any = (
  capture,
  parse,
  state
) => ({ content: parse(capture[1], state) })

const command = {
  order: 0,
  match: matchBraces,
  parse
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse
}

// String -> [Content]
const parser: (source: string) => SingleASTNode[] = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent
})

export default parser
