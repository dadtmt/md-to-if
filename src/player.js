import * as R from 'ramda'

export const getTarget = R.pipe(R.prop('target'), R.tail)

export const matchTarget = move => R.propEq('name', getTarget(move))

// To do : write a getSceneName
export const findScene = move => R.filter(matchTarget(move))

const player = (book, moves = []) => {
  const [firstScene, ...scenes] = book
  return [
    firstScene,
    ...moves.reduce((acc, move) => {
      switch (move.type) {
        case 'start':
          return [R.head(scenes)]
        case 'anchor':
          return [...acc, findScene(move)(book)]
        default:
          return acc
      }
    }, []),
  ]
}

export default player
