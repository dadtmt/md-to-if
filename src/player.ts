import playMoves, { State } from './moves'
import { Scene } from '.'

export type Move = {
  type: string
  [prop: string]: any
}
export type PlayedScene = Scene & { state: State }

// [Scene], [Move] -> [PlayedScene]
const player: (scenes: Scene[], moves?: Move[]) => Scene[] = (
  scenes,
  moves
) => {
  const [introduction, ...scenesAfterInroduction] = scenes
  return [introduction, ...playMoves(moves || [], scenesAfterInroduction)]
}

export default player
