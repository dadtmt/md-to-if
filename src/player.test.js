import parser from './parser'
import player, { findScene, getTarget, matchTarget } from './player'
import adventure from './adventure.md.js'
import book from './book'

const adventureBook = book(parser(adventure))
const startMove = {
  type: 'start',
}
const moveToCantina = {
  type: 'anchor',
  target: '/cantina',
}

describe('player', () => {
  it('play book introduction if no actions', () => {
    expect(player(adventureBook)).toMatchSnapshot()
  })
  it('play second level 2 scene on start', () => {
    const moves = [startMove]
    expect(player(adventureBook, moves)).toMatchSnapshot()
  })
  it('play cantina scene on cantina target', () => {
    const moves = [startMove, moveToCantina]
    expect(player(adventureBook, moves)).toMatchSnapshot()
  })
})

describe('findScene', () => {
  it('find a scene', () => {
    expect(findScene(moveToCantina)(adventureBook)).toMatchSnapshot()
  })
})

describe('matchTarget', () => {
  it('match cantina scene', () => {
    const cantinaScene = { content: adventureBook[2], name: 'cantina' }
    expect(matchTarget(moveToCantina)(cantinaScene)).toBe(true)
  })
  it('does not match cantina scene', () => {
    const cantinaScene = { content: adventureBook[1], name: 'the-space-ship' }
    expect(matchTarget(moveToCantina)(cantinaScene)).toBe(false)
  })
})

describe('getTarget', () => {
  it('get move target', () => {
    expect(getTarget(moveToCantina)).toBe('cantina')
  })
})
