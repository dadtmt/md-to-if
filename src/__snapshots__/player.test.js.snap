// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findScene find a scene 1`] = `Array []`;

exports[`player play book introduction if no actions 1`] = `
Array [
  Array [
    Object {
      "content": Array [
        Object {
          "content": "Choose your story",
          "type": "text",
        },
      ],
      "level": 1,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": "This is a test adventure",
          "type": "text",
        },
        Object {
          "content": ", welcome",
          "type": "text",
        },
        Object {
          "content": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
]
`;

exports[`player play cantina scene on cantina target 1`] = `
Array [
  Array [
    Object {
      "content": Array [
        Object {
          "content": "Choose your story",
          "type": "text",
        },
      ],
      "level": 1,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": "This is a test adventure",
          "type": "text",
        },
        Object {
          "content": ", welcome",
          "type": "text",
        },
        Object {
          "content": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  Array [
    Object {
      "content": Array [
        Object {
          "content": "The space ship",
          "type": "text",
        },
      ],
      "level": 2,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": "You are the brave pilot of this ship",
          "type": "text",
        },
        Object {
          "content": ".
The trip is long and boring",
          "type": "text",
        },
        Object {
          "content": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "Are you going to ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": "get food",
              "type": "text",
            },
          ],
          "target": "/cantina",
          "title": undefined,
          "type": "link",
        },
        Object {
          "content": " or to ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": "get a rest",
              "type": "text",
            },
          ],
          "target": "/bedroom",
          "title": undefined,
          "type": "link",
        },
      ],
      "type": "paragraph",
    },
  ],
  Array [],
]
`;

exports[`player play second level 2 scene on start 1`] = `
Array [
  Array [
    Object {
      "content": Array [
        Object {
          "content": "Choose your story",
          "type": "text",
        },
      ],
      "level": 1,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": "This is a test adventure",
          "type": "text",
        },
        Object {
          "content": ", welcome",
          "type": "text",
        },
        Object {
          "content": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  Array [
    Object {
      "content": Array [
        Object {
          "content": "The space ship",
          "type": "text",
        },
      ],
      "level": 2,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": "You are the brave pilot of this ship",
          "type": "text",
        },
        Object {
          "content": ".
The trip is long and boring",
          "type": "text",
        },
        Object {
          "content": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "Are you going to ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": "get food",
              "type": "text",
            },
          ],
          "target": "/cantina",
          "title": undefined,
          "type": "link",
        },
        Object {
          "content": " or to ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": "get a rest",
              "type": "text",
            },
          ],
          "target": "/bedroom",
          "title": undefined,
          "type": "link",
        },
      ],
      "type": "paragraph",
    },
  ],
]
`;
